### ISSUE TITLE
Unexpected behavior when using cameras.unproject_points world_coordinates=True

### ISSUE THREAD

**Issue by HectorAnadon** [facebookresearch/pytorch3d#658]
## 🐛 Bugs / Unexpected behaviors
I am trying to get point cloud from different views in the world coordinates. I am seeing unexpected behavior as some views are flipped. See image where green and red (View 3 (green) and 4 (red) in the code) are X flipped.
Am I missunderstanding something? Is this expected? If so, why?

Thanks for your help!

![Screenshot from 2021-04-27 13-55-36](https://user-images.githubusercontent.com/12595599/116237698-d8be9600-a760-11eb-8172-3ca9dcefc37f.png)
Obj result from the image [here](https://www.dropbox.com/s/2yketaeyn7xt5g3/github.obj?dl=0).

If I use `sample_points_from_meshes` (white points) I get the following result. So it seems that things are Y flipped.
![Screenshot from 2021-04-27 14-18-35](https://user-images.githubusercontent.com/12595599/116239995-9ba7d300-a763-11eb-835e-6a481c7712e3.png)

Code:
```
import torch
import os
from torch import nn
from pytorch3d.renderer import (
    FoVPerspectiveCameras,
    RasterizationSettings,
    PointLights,
    SoftPhongShader,
    look_at_view_transform,
    MeshRasterizer,
    look_at_rotation
)
from pytorch3d.io import load_objs_as_meshes
import numpy as np

device = 'cuda'
img_size = 224
DATA_DIR = "./data"
obj_filename = os.path.join(DATA_DIR, "cow_mesh/cow.obj")
mesh = load_objs_as_meshes([obj_filename], device=device)

class MeshRendererWithDepth(nn.Module):
    def __init__(self, rasterizer, shader):
        super().__init__()
        self.rasterizer = rasterizer
        self.shader = shader

    def forward(self, meshes_world, **kwargs) -> torch.Tensor:
        fragments = self.rasterizer(meshes_world, **kwargs)
        images = self.shader(fragments, meshes_world, **kwargs)
        return images, fragments.zbuf

raster_settings = RasterizationSettings(
    image_size=img_size,
    blur_radius=0.0,
    faces_per_pixel=1,
    perspective_correct=True
    # bin_size=0
    )

R, T = look_at_view_transform(2.7, 0, 0)
cameras = FoVPerspectiveCameras(R=R, T=T, device=device)

rasterizer = MeshRasterizer(
    raster_settings=raster_settings,
    cameras=cameras
)

shader = SoftPhongShader(
    device=device,
    cameras=cameras
)

rasterizer_depth = MeshRendererWithDepth(rasterizer=rasterizer, shader=shader)

def get_world_pointcloud(cameras):
    images, zbuf = rasterizer_depth(mesh, cameras=cameras)

    dist = torch.distributions.uniform.Uniform(-1, 1)
    grid = dist.sample((1, 1, 5000, 2)).to(device)
    depth_sample = torch.nn.functional.grid_sample(zbuf.permute(0, 3, 1, 2), grid)
    uv_depth = torch.cat([grid[0, 0], depth_sample[0, 0, 0].unsqueeze(-1)], dim=1)
    xyz_world = cameras.unproject_points(uv_depth, world_coordinates=True)

    return xyz_world

xyz_world_all = []

# View 1 - Cyan
R, T = look_at_view_transform(2.7, 0, 0)
cameras = FoVPerspectiveCameras(R=R, T=T, device=device)
xyz_world = get_world_pointcloud(cameras)
xyz_world_all.append(xyz_world)

# View 2 - Yellow
R, T = look_at_view_transform(2.7, 0, 180)
cameras = FoVPerspectiveCameras(R=R, T=T, device=device)
xyz_world = get_world_pointcloud(cameras)
xyz_world_all.append(xyz_world)

# View 3 - Red
R, T = look_at_view_transform(2.7, 0, 90)
cameras = FoVPerspectiveCameras(R=R, T=T, device=device)
xyz_world = get_world_pointcloud(cameras)
xyz_world_all.append(xyz_world)

# View 4 - Green
R, T = look_at_view_transform(2.7, 0, 270)
cameras = FoVPerspectiveCameras(R=R, T=T, device=device)
xyz_world = get_world_pointcloud(cameras)
xyz_world_all.append(xyz_world)

xyz_world_all = torch.cat(xyz_world_all)
```

Version:
```
pytorch                   1.7.1
pytorch3d                 0.4.0
```

**Comment by bottler** [comment]
This is a cool visualisation. Your two images seem to be from opposite points of view, because the order of the cyan/red/yellow/green is reversed. So in particular up on the cow is roughly down in the top image and roughly up in the bottom image. If we assume the white points are in the right place in the second image, then all the colored cows in both images are the wrong way up. A quick guess is that every colored cow has been rotated 180 degrees around the axis between its camera and the cow. Does that help you debug?

**Comment by HectorAnadon** [comment]
Thanks for your help!
Fix with `grid[0, 0]*-1`

**Comment by vaibhavnayel** [comment]
I'm trying to do the same thing as you @HectorAnadon but I'd like to avoid the poor interpolations at the object boundary which causes floating points. for this I need a mapping from pixel coordinates (0,1,...,size-1) to uv coordinates (-1,1).
I tried making a meshgrid as follows:
```
delta=1/size
u=np.linspace(-1+delta,1-delta,size)
uv_grid=np.array(np.meshgrid(u,u)).reshape(2,-1).T
```
I followed the rest of your steps and visualized the pointcloud but there appears to be a small misalignment between the projected points from different cameras. I assume this is because my mapping from pixels to uv coordinates is incorrect.

<img width="816" alt="misalignment" src="https://user-images.githubusercontent.com/19236390/120668902-38f6d500-c48f-11eb-8d6c-d830a28359ea.png">

any idea what I'm doing wrong?
